Source/bmalloc/bmalloc/DebugHeap.h

```diff
@@ -32,7 +32,7 @@
 #include <mutex>
 #include <unordered_map>
 
-#if BOS(DARWIN)
+#if BENABLE(MALLOC_HEAP_BREAKDOWN) || BOS(DARWIN)
```

Comment:
Should remove the `BOS(DARWIN)` guards here, right? The malloc.h is only wanted when heap breakdown is enabled?

```suggestion
#if BENABLE(MALLOC_HEAP_BREAKDOWN)
```

Reply:
I've checked your suggestion on intel-based mac with apple port and it breaks the compilation with default settings (no malloc heap breakdown) - apparently bmalloc's debug heap should be compiled always when `BOS(DARWIN)`

Reply:
It'd be nicer to always build everything on platforms that support it, and guard it behind a runtime setting rather than introduce a new build setting. There's nothing platform-specific in the new Malloc.h, right? It could be used everywhere and gated behind, say, a debug environment variable?

The reason I mention this: we have a long effort to reduce unnecessary build flags, and this pull request is adding a new one.

Reply:
In general, I think we should be trying to do the exact opposite; binary size is super important, and you should be able to compile out things you don't need. For linux, I think it is fine to always include this in developer mode builds, so we don't necessarily need a new flag. On macOS though, some but not all of this code is compiled in by default and we shouldn't change the defaults on that platform.

Reply:
Well, this particular feature is debug-only and meant to be used very rarely - thus my idea was to have it build-time guarded just like Apple does (see https://docs.webkit.org/Infrastructure/MemoryInspection.html). The only difference here is that the build option is exposed to meta-buildsystem.

As for the new malloc.h - the only platform specific thing about it is that Apple platforms have their own malloc.h. Other than this you're right - new malloc.h does not have anything platform specific inside.

Regarding the new build flag - you mean the cmake part right? So you'd suggest to remove that from cmakes completely and have all the new code (from non-apple platforms) gated behind MALLOC_HEAP_BREAKDOWN macro like apple does?

Reply:
> In general, I think we should be trying to do the exact opposite; binary size is super important,

The WebKit project consensus is that it's not important. WebKit is so massive that cutting a few kilobytes out is just not going to matter, even on tiny embedded devices. But we still receive many bug reports about problems caused by one build flag or another being enabled or disabled, so the cost of the build flags is real and should be avoided whenever possible.

> Regarding the new build flag - you mean the cmake part right? So you'd suggest to remove that from cmakes completely and have all the new code (from non-apple platforms) gated behind MALLOC_HEAP_BREAKDOWN macro like apple does?

I suggest putting it behind a runtime feature flag rather than a build flag. But bmalloc doesn't really have a mechanism for this other than environment variables, and I dislike environment variables. So I guess a build flag is OK. Your call.

Reply:
(If it's going to be a build flag, it's better to expose it to CMake, so that developers don't have to manually edit source code files.)

Reply:
(The usual rule is: if your feature depends on some new external dependency, add a build flag so that use of the external dependency can be disabled. Otherwise: use runtime feature checks only.)

Reply:
Well, IMO the build flag also resonates better with current implementation by Apple as `MALLOC_HEAP_BREAKDOWN` macro is already used by Apple ports so by exposing it to cmake as build option, we make Apple's devs lives easier as a side effect.

In that case IMO the current state of PR is what we need.

Reply:
There is value in matching what already exists, sure.